Cleaned up ReactHardwareComponent

edvinerikson opened the pull request:
* checks if the native component is supported, warns otherwise.  
  `<foo />` will warn. `<pin />` will not.
  Currently only `pin` and `container` is available
  Not sure if I should check in `receiveComponent` too.
* Added different view configs for the components.
* cleaned up the test a little bit.
* removed the possibility to use any native component name.  
  e.g `<foo />` and `<pin />` would act the same, but not anymore.

edvinerikson commented on src/ReactHardwareComponent.js:
```diff
@@ -42,16 +64,23 @@ type ViewConfigPropTypes = {
+  this._currentElement = element;
+  this.viewConfig = DEFAULT_VIEW_CONFIG;
+  if (typeof element.type === 'string') {
+    this.viewConfig = getViewConfig(element.type);
```
Not sure if should run this once again in `receiveComponent`

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -21,14 +20,37 @@ export type ReactHardwareComponentViewConfig = {
 
 export const DEFAULT_VIEW_CONFIG = {
   validAttributes: {
+  },
```
I think that the PIN_VIEW_CONFIG should still be the `DEFAULT_VIEW_CONFIG`. I suspect that as we start heading down this path that the default case will still be the same attributes as a `pin`.

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -21,14 +20,37 @@ export type ReactHardwareComponentViewConfig = {
+};
+
+export const CONTAINER_VIEW_CONFIG = {
+  validAttributes: {
```
nit: make the empty object a single line: `validAttributes: {},`

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -42,16 +64,23 @@ type ViewConfigPropTypes = {
+    this.viewConfig = getViewConfig(element.type);
+  }
+  if (process.env.NODE_ENV !== 'production') {
+    invariant(
```
ReactDOM when given something like `<foobar />` can still create and render it. Do you happen to know what RN does off hand?

Currently I’d rather change this to a `warning` and just revert the type to a `pin`/use the `PIN_VIEW_CONFIG` with `UnknownComponent` as the `uiViewClassName` to not throw an error in the rendering.

edvinerikson replied:
This is what RN does: https://github.com/facebook/react/blob/master/src/renderers/native/ReactNative/ReactNativeDefaultInjection.js#L101-L108

A warning sounds good. I am not sure that defaulting to a pin is a good idea. I'd rather just warn and try to render (nothing will happen because we only call `setPayload` when it is a pin). This way people will notice that they are using an invalid type.

iamdustan replied:
that may be more okay on RN where the thrown error happens on the JS thread and won’t (maybe won’t?) change the hardware output. If you’re rendering into hardware and an error gets thrown in the middle of some updates we may leave the hardware in an invalid/undesirable state. This is me “trying to do the right thing” in that scenario.

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -42,16 +64,23 @@ type ViewConfigPropTypes = {
+  if (process.env.NODE_ENV !== 'production') {
+    invariant(
+      element.type === 'pin' || element.type === 'container',
+      '%s is not a defined component',
```
`Attempted to render an unsupported generic component "%s". Must be "pin" or "container". Check the render method of %s`

You can see how to do this at https://github.com/facebook/react/blob/eb85b7b2c989faeb6ed8cd787cf67bd5f6d3ec7d/src/renderers/dom/client/ReactMount.js#L319-L321

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -69,12 +98,6 @@ ReactHardwareComponent.Mixin = {
     return this;
   },
 
-  // TODO: React 0.15 began removing construct calls and inlining directly into
```
:thumbsup:

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -92,45 +115,40 @@ ReactHardwareComponent.Mixin = {
-          'attribute is immutable. Check the render function of ' +
-          nextElement.displayName  + '.' // TODO
+    // special cases
+    switch (nextElement.type) {
```
for now since we only have one `case` I’d like to revert this to a simple `if (nextElement.type === 'pin')`

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -145,48 +163,34 @@ ReactHardwareComponent.Mixin = {
-    if (payload) {
-      if (process.env.NODE_ENV !== 'production') {
-        HardwareManager.validatePayloadForPin(rootID, payload);
+    switch (this._currentElement.type) {
```
same here, change this to `if (this._currentElement.type === 'pin')`

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -145,48 +163,34 @@ ReactHardwareComponent.Mixin = {
+        // https://github.com/facebook/react-native/blob/9f48c004ba866aa24d17242a817929462a091179/Libraries/ReactNative/ReactNativeBaseComponent.js
+        //
+        // ReactDOM _createInitialChildren(transaction, props, context, lazyTree);
+        //   https://github.com/facebook/react/blob/master/src/renderers/dom/shared/ReactDOMComponent.js#L725
```
while cleaning up this PR this entire comment block could actually be removed :smile:

iamdustan commented on src/__tests__/ReactHardwareComponent-test.js:
```diff
@@ -4,6 +4,11 @@ import ReactHardwareComponent from '../ReactHardwareComponent';
 describe('ReactHardwareComponent', () => {
+  it('should warn if it gets an unknown type', () => {
+    // TODO: check so it is the correct error
+    expect(() => new ReactHardwareComponent({ type: 'foo' })).toThrow();
```
nit: remove outter spacing in single line object literals (following React code style) `{type: 'foo'}`

iamdustan commented on src/__tests__/ReactHardwareComponent-test.js:
```diff
@@ -4,6 +4,11 @@ import ReactHardwareComponent from '../ReactHardwareComponent';
 
 describe('ReactHardwareComponent', () => {
+  it('should warn if it gets an unknown type', () => {
+    // TODO: check so it is the correct error
```
based on the previous comment this should be a warning so you’ll spyOn and `console.warn`

iamdustan commented on src/__tests__/ReactHardwareComponent-test.js:
```diff
@@ -25,55 +30,45 @@ describe('ReactHardwareComponent', () => {
-          value: true,
-        },
+  describe('pin', () => {
+    describe('receiveComponent', function() {
```
nit: while you’re in here could you update these `function() {`s to `() => {`? no issue either way

iamdustan commented on src/__tests__/ReactHardwareComponent-test.js:
```diff
@@ -25,55 +30,45 @@ describe('ReactHardwareComponent', () => {
-          pin: true,
-          value: true,
-        },
+  describe('pin', () => {
```
do we plan on attending tests for more than just the `pin` right now? If not I’m okay leaving this in the top-level `describe` and minimizing the diff.

edvinerikson replied:
I don't think so. I just added it that way because the tests only tested the pin component. 
I'll revert it back. :+1:

iamdustan commented on src/components/Container.js:
```diff
@@ -10,14 +10,10 @@
   render() {
     return (
-      <container {...containerProp}>
+      <container>
```
🙌 nice work removing my hack!

iamdustan commented:
:tada: yay! You win the first contribution to React Hardware award!

Solid direction on these changes. A few tweaks (feel free to push back on them as well) and we’ll merge it in!

edvinerikson commented:
I have updated the PR based on your feedback (thanks :D). Notably changes are the viewConfig, a eslint rule that I added and a `checkRenderMethod` function.

The viewConfig now have a `EMPTY_VIEW_CONFIG` and a `GENERIC_VIEW_CONFIG` instead of a `DEAFULT_VIEW_CONFIG`. I did this because I think it is wrong to pass the pin/generic config into everything we don't recognize. Right now I give anything that we don't recognize the `EMPTY_VIEW_CONFIG` which I think is a better approach.

The linting rule I added enforces object-curly-spaces to never occur (gives an error). 

The reason I added the `checkRenderMethod` function is because in some cases we might not have a owner to reference to and if that is the case we don't have any reason to include that in the final message.

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -29,6 +30,27 @@ export const DEFAULT_VIEW_CONFIG = {
+  if (type === 'pin') {
+    return GENERIC_VIEW_CONFIG;
+  }
+  return EMPTY_VIEW_CONFIG;
```
is `container` considered an `EMPTY_VIEW_CONFIG` then?

edvinerikson replied:
Yes

iamdustan commented on src/ReactHardwareComponent.js:
```diff
@@ -29,6 +30,27 @@ export const DEFAULT_VIEW_CONFIG = {
+
+// In some cases we might not have a owner and when
+// that happens there is no need to inlcude "Check the render method of ...".
+function checkRenderMethodMessage() {
```
nit:

```jsx
const checkRenderMethod = () => ReactCurrentOwner.owner && ReactCurrentOwner.owner.getName()
  ? ` Check the render method of `${ReactCurrentOwner.owner.getName()}.` : '';
```

edvinerikson replied:
:+1: The first one followed the React style, but I like this one more. :)

iamdustan replied:
es6 ftw!

iamdustan commented:
last nit, few new eslint errors to fix from that rule change, squash to one commit, and merging!

edvinerikson commented:
Squashed and done. :+1:
